espwrap/adaptors/sendgrid_v3.py

```diff
@@ -0,0 +1,201 @@
+                message.substitution = Substitution(new_key, val)
+
+            #Content (Text)
+            if self.body[MIMETYPE_TEXT]:
+                content = Content(
+                    MimeType.text,
+                    self.body[MIMETYPE_TEXT]
+                )
+                message.content = content
+
+            #Content (Html)
+            if self.body[MIMETYPE_HTML]:
+                content = Content(
+                    MimeType.html,
+                    self.body[MIMETYPE_HTML]
+                )
+                message.content = content
+                
+            #Attachment
+            if self.attachments:
+                for file_name, file_path_or_string in self.attachments.iteritems():
+                    attachment = Attachment()
+                    encoded = base64.b64encode(str(file_path_or_string))
+                    attachment.file_content = FileContent(encoded)
+                    attachment.file_name = FileName(file_name)
+                message.attachment = attachment
+
+            #Webhook Data
+            if self.webhook_data:
+                for key, val in self.webhook_data.items():
+                    message.custom_arg = CustomArg(key, val)
+            if self.template_name:
+                message.custom_arg = CustomArg('template_name', self.template_name)
+                    
+            #Tracking
+            tracking_settings = TrackingSettings()            
+            #Opens
+            if self.track_opens:
+                tracking_settings.open_tracking = OpenTracking(
+                    self.track_opens,
+                    OpenTrackingSubstitutionTag("open_tracking")
+                )
+            #Clicks
+            if self.track_clicks:
+                tracking_settings.click_tracking = ClickTracking(
+                    self.track_clicks,
+                    self.track_clicks
+                )
+            message.tracking_settings = tracking_settings
+            
+            #Personalizations (DO THIS LAST)
+            for subgrp in to_send:
+                p=0
+                for recip in subgrp:
+                    message.from_email = From(self.from_addr)
+                    message.to = To(recip['email'], p=p)
+                    message.subject = Subject(self.subject)
+
+                    for key, val in recip['merge_vars'].items():
+                        new_key = '{1}{0}{2}'.format(key, *self.delimiters)
+                        message.substitution = Substitution(new_key, val, p=p)
+                    p+=1
+
+            try:
+                response = self.client.send(message)
+                print("status code")
+                print(response.status_code)
+                print("body")
+                print(response.body)
+                print("headers")
+                print(response.headers)
+            except exceptions.BadRequestsError as e:
+                print(e.body)
+                exit()
```

Comment:
I'd replace exit with raise. You generally don't want libraries to ever throw exit codes. It's up to the person using your library to handle this error situation.

Reply:
agreed and replaced. i borrowed this block from sendgrid repo use cases.
now uses raise. will evolve from here ... not sure what other exceptions need to be addressed, but the plan is to at least manage sendgrid API limitations and log when the payload is too big, too many personalizations, etc.